druid/src/core.rs

```diff
@@ -559,160 +559,173 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
 
-        // Note: could also represent this as `Option<Event>`.
-        let mut recurse = true;
-        let child_event = match event {
+        let recurse = match event {
             Event::Internal(internal) => match internal {
                 InternalEvent::MouseLeave => {
                     let hot_changed = WidgetPod::set_hot_state(
                         &mut self.inner,
-                        child_ctx.widget_state,
-                        child_ctx.state,
+                        &mut self.state,
+                        ctx.state,
                         rect,
                         None,
                         data,
                         env,
                     );
-                    recurse = had_active || hot_changed;
-                    Event::Internal(InternalEvent::MouseLeave)
+                    needs_merge |= hot_changed;
+                    had_active || hot_changed
                 }
                 InternalEvent::TargetedCommand(target, cmd) => {
                     match target {
-                        Target::Window(_) => Event::Command(cmd.clone()),
-                        Target::Widget(id) if *id == child_ctx.widget_id() => {
-                            Event::Command(cmd.clone())
+                        Target::Widget(id) if *id == self.id() => {
+                            is_handled = true; // Prevent it being passed down to children
+                            modified_event = Some(Event::Command(cmd.clone()));
+                            true
                         }
                         Target::Widget(id) => {
                             // Recurse when the target widget could be our descendant.
                             // The bloom filter we're checking can return false positives.
-                            recurse = child_ctx.widget_state.children.may_contain(id);
-                            Event::Internal(InternalEvent::TargetedCommand(*target, cmd.clone()))
+                            self.state.children.may_contain(id)
                         }
-                        Target::Global => {
-                            panic!("Target::Global should be converted before WidgetPod")
+                        Target::Global | Target::Window(_) => {
+                            modified_event = Some(Event::Command(cmd.clone()));
+                            true
                         }
                     }
                 }
                 InternalEvent::RouteTimer(token, widget_id) => {
-                    let widget_id = *widget_id;
-                    if widget_id != child_ctx.widget_state.id {
-                        recurse = child_ctx.widget_state.children.may_contain(&widget_id);
-                        Event::Internal(InternalEvent::RouteTimer(*token, widget_id))
+                    if *widget_id == self.id() {
+                        modified_event = Some(Event::Timer(*token));
+                        true
```

Comment:
So, I think I got this now:
Only commands targeted the current widget are marked as handled and will thus be ignored by nested `WidgetPod`s while Timer events are filtered out regardless of being handled, as they are always targeted at a single widget.

This distinction is kind of irritating when reading the code, and should be documented if we stick to it.

Also I think it is weird that a widget, receiving a command targeted at it, gets that command already marked as `handled`. I do understand that this is currently necessary because we loose the information about whether or not the command was targeted, once we convert it to a normal `Event::Command`, but it still seems wrong to me.

The two possible solutions that come to my mind are either passing down both, the normal command and the internal in case it should propagate further, or adding the `Target` to `Event::Command` and always pass it down along the command.

Reply:
Initially I tried solving it by passing down the internal command, however that breaks a whole class of command usage. Stuff like `ALL_TAKE_FOCUS_AFTER` in the test case. A command where a widget wants to decide whether to handle it after it has been passed down to children. This would no longer work if we pass an internal command that generates the regular command for each widget.

Adding `Target` to `Event::Command` is also something I considered, but I went with `is_handled` because that is backwards compatible. I'm not commited to it though, and I'll write an extra comment with the pros and cons of doing either.